app/util/event_log.py

```diff
@@ -0,0 +1,265 @@
+
+        :param tag: A string identifier that describes the event being logged (e.g., "REQUEST_SENT")
+        :type tag: str
+        :param log_msg: A message that will also be logged to the human-readable log (not the event log). It will be
+            string formatted with tthe event_data dict.  This is a convenience for logging to both human- and
+            machin-readable logs.
+        :type log_msg: str
+        :param event_data: Free-form key value pairs that make up the event
+        :type event_data: dict
+        """
+        if not self.logging_disabled:
+            event_data['__id__'] = self._generate_event_id()
+            event_data['__tag__'] = tag
+            event_data['__timestamp__'] = time.time()
+            json_dumps = json.dumps(event_data, sort_keys=True)
+            self._write_to_analytics_logger(json_dumps)
+            self._write_to_log_cache(event_data)
+
+        if log_msg:
+            self._logger.info(log_msg, **event_data)
+
+    def _write_to_analytics_logger(self, json_dumps):
+        """
+        :param json_dumps: A json encoded message that describes the event being logged
+        :type json_dumps: str
+        """
+        if self._analytics_logger:
+            self._analytics_logger.event(json_dumps)  # pylint: disable=no-member
+
+    def _write_to_log_cache(self, event_data):
+        """
+        :param event_data: Free-form key value pairs that make up the event
+        :type event_data: dict
+        """
+        self._log_cache.append(event_data)
+        self._expire_stale_items_in_cache()
+
+    def _expire_stale_items_in_cache(self):
+        """
+        Expires stale items in the cache. An item in the cache is considered stale if the timestamp is
+        significantly older than the most recent item. This expiration time is determined by the constant,
+        LOG_CACHE_EXPIRE_TIME_IN_HOURS.
+
+        Since we do not want to artificially impose limits on the log cache, we will not expire items
+        if the total length of the log cache is less than LOG_CACHE_THRESHOLD.
+        """
+        while len(self._log_cache) > LOG_CACHE_THRESHOLD and self.oldest_cache_event_is_stale():
+            self._log_cache.popleft()
+
+    def oldest_cache_event_is_stale(self):
+        """
+        Returns whether the time stamp of the event data is stale in comparison with the start time.
+        The time is considered stale if it is is older than start_time + LOG_CACHE_EXPIRE_TIME_IN_HOURS
+
+        :rtype: bool
+        """
+        if len(self._log_cache) == 0:
+            return False
+        else:
+            latest = self._latest_timestamp_in_cache()
+            oldest = self._oldest_timestamp_in_cache()
+            return oldest + LOG_CACHE_EXPIRE_TIME_IN_HOURS * 60 * 60 < latest
+
+    def _latest_timestamp_in_cache(self):
+        """
+        :rtype: float | None
+        """
+        if len(self._log_cache) == 0:
+            return None
```

Comment:
Ah, I see that currently hitting that path would be impossible since you do the length == 0 check in all three places. Do you think it might be slightly more robust if instead of doing the length == 0 check on line 150 if we just caught the TypeError and returned False? (To make it more explicit what we are guarding against?)